### PR TITLE
Jaeger tracing - Add tracing framework and samples

### PR DESCRIPTION
**Pull request by tyacbovi**
The following features work whenever `enable_tracing` is set to `true` and
`jaeger_server_ip` is also set with the corresponding Jaeger API Server
IP.
- Adds automatic tracing to thread context whenever an API call is made
- Adds a decorator for tracing, it starts a new span before the function
is called.

### CODE REVIEW

File: rest-service/manager_rest/storage/models_base.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
 from sqlalchemy.ext.hybrid import HYBRID_PROPERTY
 from sqlalchemy.orm.interfaces import NOT_EXTENSION
 
-from manager_rest.utils import classproperty
+from manager_rest.utils import classproperty, with_tracing
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/models_base.py]: not in use

File: rest-service/manager_rest/storage/storage_manager.py [hunk 2]
```diff
@@ -36,7 +40,12 @@
 
 
 class SQLStorageManager(object):
+    @with_tracing()
+    def __init__(self):
+        self._enable_tracing()
+
     @staticmethod
+    @with_tracing()
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: why not use the wrapper?

File: rest-service/manager_rest/storage/storage_manager.py [hunk 3]
```diff
@@ -605,7 +643,13 @@ def refresh(self, instance):
         self._load_relationships(instance)
         return instance
 
+    @with_tracing()
+    def _enable_tracing(self):
+        """Enable tracing for all SQLAlchemy engines.
+        """
+        SQLAlchemyTracingHelper.enable_tracing_for_all_engines()
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: why there is a need for this and for the decorator

File: rest-service/manager_rest/storage/storage_manager.py [hunk 4]
```diff
@@ -629,3 +673,95 @@ def __iter__(self):
...
+        Clear an object's decorated tracing fields,
+        to prevent unintended further tracing.
+        '''
+        if hasattr(obj, '_parent_span'):
+            del obj._parent_span
+        if hasattr(obj, '_traced'):
+            del obj._traced
+
+    @staticmethod
+    def _normalize_stmt(statement):
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: Explain what is normal and why this is needed?

File: rest-service/manager_rest/storage/storage_manager.py [hunk 5]
```diff
@@ -629,3 +673,95 @@ def __iter__(self):
...
+    def _engine_before_cursor_handler(conn, cursor,
+                                      statement, parameters,
+                                      context, executemany):
+        current_span = get_current_span()
+        if not current_span:
+            return
+
+        stmt_obj = None
+        if context.compiled is not None:
+            stmt_obj = context.compiled.statement
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: Those three lines can be reduced to a single line

File: rest-service/manager_rest/storage/storage_manager.py [hunk 6]
```diff
@@ -629,3 +673,95 @@ def __iter__(self):
...
+
+        # Don't trace PRAGMA statements coming from SQLite
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: why not?

File: rest-service/manager_rest/storage/storage_manager.py [hunk 7]
```diff
@@ -629,3 +673,95 @@ def __iter__(self):
...
+    def _engine_after_cursor_handler(conn, cursor,
+                                     statement, parameters,
+                                     context, executemany):
+        span = getattr(context, '_span', None)
+        if span is None:
+            return
+
+        span.finish()
+
+        if context.compiled is not None:
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: Add why this is needed

File: rest-service/manager_rest/utils.py [hunk 8]
```diff
@@ -275,3 +278,33 @@ def send_event(event, message_type):
...
     events_publisher.close()
+
+
+def with_tracing(func_name=None):
+    """Wrapper function for Flask operation call tracing.
+    This decorator must be activate iff the following condition apply. One of
+    the parent calls must be done inside a 'with span_in_context(span)' scope
+    (otherwise you'd get a parentless span).
+
+    :param name: name to display in tracing
```
Comment (tyacbovi) [inline rest-service/manager_rest/utils.py]: Not clear the variable name and explanation

Comment (tyacbovi) [reply rest-service/manager_rest/utils.py]: and the variable name had change

File: rest-service/manager_rest/storage/storage_manager.py (hunk 2 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: `sqlalchemy` wrapper? Because it doesn't use `threading.local()` which what I implemented is based on.

File: rest-service/manager_rest/storage/storage_manager.py (hunk 3 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: This enables SQL Engines tracing. Otherwise you won't see any queries.

File: rest-service/manager_rest/storage/storage_manager.py (hunk 4 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: I looked over the implementation of the [sqlalchemy wrapper](https://github.com/opentracing-contrib/python-sqlalchemy/blob/master/sqlalchemy_opentracing/__init__.py#L138) and they had it. I guess this is because otherwise you'd get a bloated text when you view the query.

File: rest-service/manager_rest/storage/storage_manager.py (hunk 6 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: Thought they'd be many of them but we're here to trace. Removed it (it was copied from the `sqlalchemy` opentracing wrapper).

File: rest-service/manager_rest/storage/storage_manager.py (hunk 7 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: Forgot to remove these leftovers. Removed.

File: rest-service/manager_rest/storage/storage_manager.py (hunk 5 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: Done.

File: rest-service/manager_rest/storage/storage_manager.py [hunk 9]
```diff
@@ -629,3 +673,95 @@ def __iter__(self):
 
     def __getitem__(self, item):
         return self.items[item]
+
+
+class SQLAlchemyTracingHelper(object):
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: Done

File: rest-service/manager_rest/server.py [hunk 10]
```diff
@@ -149,6 +176,31 @@ def handle_exception(flask_method, flask_restful_method, e):
...
         self.config['ERROR_404_HELP'] = False
 
+    def dispatch_request(self):
+        """Wraps up the super 'dispatch_request' func to enable easier tracing.
+        """
+        if not self.tracer:
+            return super(CloudifyFlaskApp, self).dispatch_request()
+
+        request = stack.top.request
+        operation_name = '{} ({})'.format(str(request.endpoint),
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/server.py]: Done

File: rest-service/manager_rest/storage/storage_manager.py [hunk 11]
```diff
@@ -629,3 +673,95 @@ def __iter__(self):
...
+    @staticmethod
+    def _engine_error_handler(exception_context):
+        execution_context = exception_context.execution_context
+        span = getattr(execution_context, '_span', None)
+        if span is None:
+            return
+
+        exc = exception_context.original_exception
+        span.set_tag('sqlalchemy.exception', str(exc))
+        span.set_tag('error', 'true')
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: It can be any value. `jaeger_client` parses it to string anyhow. I copied this from another place. Changed.

File: rest-service/manager_rest/storage/storage_manager.py (hunk 4 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: I reviewed this once again and I think without normalization it's much more readable

File: rest-service/manager_rest/server.py [hunk 12]
```diff
@@ -88,6 +92,29 @@ def __init__(self, load_config=True):
         setup_resources(Api(self))
         self.register_blueprint(app_errors)
 
+        self.tracer = None
+        if config.instance.enable_tracing:
+            self.before_first_request(self._init_jeager_tracer)
+
+    def _init_jeager_tracer(self):
```
Comment (tyacbovi) [inline rest-service/manager_rest/server.py]: misspell 'jaeger'

File: rest-service/manager_rest/security/secured_resource.py [hunk 13]
```diff
@@ -18,14 +18,15 @@
...
-from manager_rest.utils import abort_error
+from manager_rest.utils import abort_error, traces
 from manager_rest.manager_exceptions import MissingPremiumPackage
 
 from .authentication import authenticator
 
 
 def authenticate(func):
     @wraps(func)
+    @traces('authenticate')
```
Comment (tyacbovi) [inline rest-service/manager_rest/security/secured_resource.py]: This should work if your wrap 'authenticate' func instead?

File: rest-service/manager_rest/server.py [hunk 14]
```diff
@@ -88,6 +92,29 @@ def __init__(self, load_config=True):
         setup_resources(Api(self))
         self.register_blueprint(app_errors)
 
+        self.tracer = None
+        if config.instance.enable_tracing:
```
Comment (tyacbovi) [inline rest-service/manager_rest/server.py]: Add short explanation why this is needed

File: rest-service/manager_rest/server.py [hunk 15]
```diff
@@ -149,6 +176,29 @@ def handle_exception(flask_method, flask_restful_method, e):
             flask_restful_handle_user_exception)
         self.config['ERROR_404_HELP'] = False
 
+    def dispatch_request(self):
+        """Wraps up the super 'dispatch_request' func to enable easier tracing.
```
Comment (tyacbovi) [inline rest-service/manager_rest/server.py]: Add more about what this func do

File: rest-service/manager_rest/storage/storage_manager.py [hunk 16]
```diff
@@ -36,7 +40,12 @@
 
 
 class SQLStorageManager(object):
+    @traces()
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: Why do u need to use 'traces' when u are init-ing it?

File: rest-service/manager_rest/storage/storage_manager.py [hunk 17]
```diff
@@ -605,7 +643,14 @@ def refresh(self, instance):
         self._load_relationships(instance)
         return instance
 
+    @traces()
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: This will work when this func was not called before?

File: rest-service/manager_rest/storage/storage_manager.py [hunk 18]
```diff
@@ -629,3 +674,59 @@ def __iter__(self):
 
     def __getitem__(self, item):
         return self.items[item]
+
+
+def enable_tracing_for_all_engines():
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: maybe export all of this functions to a utils file for tracing

File: rest-service/manager_rest/test/endpoints/test_app.py [hunk 19]
```diff
@@ -29,3 +32,75 @@ def test_get_root_404(self):
...
+
+@attr(client_min_version=1, client_max_version=base_test.LATEST_API_VERSION)
+class AppWithTracingNoIPTestCase(base_test.BaseServerTestCase):
+    """Test the basic setup of the tracer and dispatch wrapping for the
+    tracing when tracing is enabled but no endpoint IP is set.
+    """
+
+    def setUp(self):
+        super(AppWithTracingNoIPTestCase, self).setUp()
+        # Dummy call to invoke the tracer initialization if enabled.
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_app.py]: Why is this needed?

File: rest-service/manager_rest/test/endpoints/test_app.py [hunk 20]
```diff
@@ -29,3 +32,75 @@ def test_get_root_404(self):
...
+        self.app.get('/')
+
+    def create_configuration(self):
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_app.py]: it is repeating itself needs to go to a public file or in the base test case class

File: rest-service/manager_rest/test/endpoints/test_app.py [hunk 21]
```diff
@@ -29,3 +32,75 @@ def test_get_root_404(self):
...
+            AppWithTracingNoIPTestCase, self).create_configuration()
+        test_config.enable_tracing = True
+        test_config.tracing_endpoint_ip = None
+        return test_config
+
+    def test_does_not_init_tracer_no_tracer_ip(self):
+        self.jaeger_mock_config.assert_not_called()
+
+
+@attr(client_min_version=1, client_max_version=base_test.LATEST_API_VERSION)
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_app.py]: you can fix this to latest to latest

File: rest-service/manager_rest/test/endpoints/test_app.py [hunk 22]
```diff
@@ -29,3 +32,75 @@ def test_get_root_404(self):
...
+                    'reporting_host': self._tracing_endpoint_ip},
+                'logging': True
+            },
+            service_name='cloudify-manager'
+        )
+        self.jaeger_mock_config.initialize_tracer.assert_called()
+
+    def test_adds_error_to_tags(self):
+        e = UnsupportedFormatException()
+        kwargs = {'tags': {"Extract failed": str(e)},
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_app.py]: better name is needed

File: rest-service/manager_rest/test/endpoints/test_utils.py [hunk 23]
```diff
@@ -95,6 +97,104 @@ def mock_get_platform():
...
+        # Dummy call to invoke the tracer initialization if enabled.
+        self.app.get('/')
+
+        foo = MagicMock(return_value='something')
+        foo.__name__ = 'foo'
+        app_patcher = patch('flask.current_app')
+        curr_span_patcher = patch(
+            'opentracing_instrumentation.request_context.get_current_span')
+        span_ctx_patcher = patch(
+            'opentracing_instrumentation.request_context.span_in_context')
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_utils.py]: look at using patch as decorator

File: rest-service/manager_rest/test/endpoints/test_utils.py [hunk 24]
```diff
@@ -95,6 +97,104 @@ def mock_get_platform():
                 self.assertTrue(
                     plugin_installable_on_current_platform(plugin))
 
+    def test_tracing_wrapper_does_nothing(self):
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_utils.py]: Not exactly right about what it tests

File: rest-service/manager_rest/test/endpoints/test_utils.py [hunk 25]
```diff
@@ -95,6 +97,104 @@ def mock_get_platform():
...
+        curr_span_patcher.stop()
+        span_ctx_patcher.stop()
+
+
+class TestUtilsWithTracingTestCase(base_test.BaseServerTestCase):
+    """Test the tracing wrapper when tracing is enabled.
+    """
+    _tracing_endpoint_ip = 'some_ip'
+
+    def setUp(self):
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_utils.py]: remove commented code

File: rest-service/manager_rest/test/endpoints/test_utils.py [hunk 26]
```diff
@@ -95,6 +97,104 @@ def mock_get_platform():
...
+
+        app_patcher.stop()
+        curr_span_patcher.stop()
+        span_ctx_patcher.stop()
+
+
+class TestUtilsWithTracingTestCase(base_test.BaseServerTestCase):
+    """Test the tracing wrapper when tracing is enabled.
+    """
+    _tracing_endpoint_ip = 'some_ip'
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_utils.py]: maybe create a test class of yourself

File: rest-service/manager_rest/security/secured_resource.py (hunk 13 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/security/secured_resource.py]: Then the decoration is performed out of context and therefore it won't have access to the `tracer` obj

File: rest-service/manager_rest/storage/storage_manager.py (hunk 16 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: It is a function that is being called, so thought it's something useful. Should I just remove it then?

File: rest-service/manager_rest/storage/storage_manager.py (hunk 17 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: I don't understand the question or more specifically why it's possible for it _not_ to work?

File: rest-service/manager_rest/storage/storage_manager.py (hunk 18 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: This only relates to SQL stuff, you're saure?

File: rest-service/manager_rest/test/endpoints/test_app.py (hunk 19 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/test/endpoints/test_app.py]: Tracer is initialized before the first request comes in and only then. Therefore I need a request to invoke that.
Why can't I initialize it normally? Because if I do so in the `__init__` or I presume, in the thread that runs the `__init__` of the Flask app, then it hangs. I can explain my theory further if needed.

File: rest-service/manager_rest/test/endpoints/test_app.py (hunk 20 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/test/endpoints/test_app.py]: How would I create different setups then? I don't think there's a different option here.

File: rest-service/manager_rest/test/endpoints/test_app.py (hunk 22 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/test/endpoints/test_app.py]: Not sure if my current solution is better. I changed it to `start_span_kwargs`. Is that better? Or not what you meant?

File: rest-service/manager_rest/test/endpoints/test_utils.py (hunk 26 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/test/endpoints/test_utils.py]: I kinda disagree with creating a new test class since the changes made to `BaseServerTestCase` were not that big. Also, I need a different setup therefore I don't think it'll save any work here.

Comment (tyacbovi) [reply rest-service/manager_rest/test/endpoints/test_utils.py]: There too much copied code

File: rest-service/manager_rest/test/endpoints/test_utils.py [hunk 27]
```diff
@@ -95,6 +97,78 @@ def mock_get_platform():
                 self.assertTrue(
                     plugin_installable_on_current_platform(plugin))
 
+    @patch('flask.current_app')
+    @patch('opentracing_instrumentation.request_context.span_in_context')
+    @patch('opentracing_instrumentation.request_context.get_current_span')
+    def test_tracing_wrapper_returns_original_func(self, curr_span, _,
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_utils.py]: does not describe correctly

File: rest-service/manager_rest/test/endpoints/test_utils.py [hunk 28]
```diff
@@ -95,6 +97,78 @@ def mock_get_platform():
...
+    @patch('manager_rest.utils.get_current_span')
+    def test_tracing_wrapper_works(self, curr_span, span_ctx):
+        foo = MagicMock(return_value='something')
+        foo.__name__ = 'foo'
+
+        curr_span.return_value = curr_span
+        current_app.tracer.start_span.reset_mock()
+
+        f = traces('not_foo')(foo)
+        r = f('bar', k='v')
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_utils.py]: this fake func is returning several times, so make it shared

File: rest-service/manager_rest/utils.py [hunk 29]
```diff
@@ -275,3 +278,32 @@ def send_event(event, message_type):
...
+        @wraps(f)
+        def with_tracing_wrapper(*args, **kwargs):
+            tracer = getattr(current_app, 'tracer', None)
+            if not tracer:
+                return f(*args, **kwargs)
+            root_span = get_current_span()
+            with tracer.start_span(name, child_of=root_span) as span:
+                with span_in_context(span):
+                    return f(*args, **kwargs)
+
```
Comment (tyacbovi) [inline rest-service/manager_rest/utils.py]: lose the not needed space

File: rest-service/manager_rest/utils.py [hunk 30]
```diff
@@ -275,3 +278,32 @@ def send_event(event, message_type):
...
+        return with_tracing_wrapper
+
```
Comment (tyacbovi) [inline rest-service/manager_rest/utils.py]: in here too

File: rest-service/manager_rest/storage/storage_manager.py (hunk 16 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: I'll rename the enable_tracing function, that enables tracing of the SQL not Python execution tracing.

File: rest-service/manager_rest/security/secured_resource.py (hunk 13 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/security/secured_resource.py]: We can trace two things here, either when the decorator was called (it'll be out of flask app context) _or_ trace when the inner function runs (I think we should be tracing the inner function).
To trace the inner function execution - the function that executes every time the wrapped function is executed, I'd have to place the @traces decorator where it's at.

File: rest-service/manager_rest/test/endpoints/test_utils.py (hunk 26 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/test/endpoints/test_utils.py]: I'll need help with this one then

File: rest-service/manager_rest/rest/rest_decorators.py [hunk 31]
```diff
@@ -91,6 +92,7 @@ def wrapper(*args, **kwargs):
 
 def exceptions_handled(func):
     @wraps(func)
+    @traces('exceptions_handled')
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/rest/rest_decorators.py]: Same reason as above decorator

File: rest-service/manager_rest/rest/rest_decorators.py [hunk 32]
```diff
@@ -123,6 +125,7 @@ class must have a "resource_fields" class variable
 
     def __call__(self, f):
         @wraps(f)
+        @traces('marshal_with {}'.format(self.response_class.__name__))
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/rest/rest_decorators.py]: Every time the interpreter decorates the function. I think only once since it's imported at the endpoint mapper (via import of the rest handlers classes)

File: rest-service/manager_rest/utils.py (hunk 30 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/utils.py]: That's a PEP8 thing, newline at the end of files.

File: rest-service/manager_rest/storage/storage_manager.py (hunk 16 above)
Comment (davidginzbourg) [reply rest-service/manager_rest/storage/storage_manager.py]: Is that OK?

File: rest-service/manager_rest/rest/rest_decorators.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/rest/rest_decorators.py]: Is that OK then?

File: rest-service/manager_rest/config.py [hunk 33]
```diff
@@ -16,6 +16,7 @@
 import os
 import yaml
 
+import jaeger_client as j_client
```
Comment (tyacbovi) [inline rest-service/manager_rest/config.py]: I missed this, there is a confusion here yaml is python standard lib and json is

File: rest-service/manager_rest/config.py [hunk 34]
```diff
@@ -94,6 +100,20 @@ def load_from_file(self, filename, namespace=''):
                     "Ignoring unknown key '{0}' in configuration file "
                     "'{1}'".format(key, filename))
 
+    def _create_tracer_config(self):
+        """Creates the tracer config.
+        """
```
Comment (tyacbovi) [inline rest-service/manager_rest/config.py]: This comment does not add info, so maybe remove it

File: rest-service/manager_rest/storage/storage_manager.py [hunk 35]
```diff
@@ -629,3 +674,59 @@ def __iter__(self):
...
+    listen(Engine, 'before_cursor_execute', _engine_before_cursor_handler)
+    listen(Engine, 'after_cursor_execute', _engine_after_cursor_handler)
+    listen(Engine, 'handle_error', _engine_error_handler)
+
+
+def _engine_before_cursor_handler(conn, cursor, statement, parameters, context,
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: maybe enter these functions to the function's scope

File: rest-service/manager_rest/storage/storage_manager.py [hunk 36]
```diff
@@ -629,3 +674,59 @@ def __iter__(self):
...
+                                  executemany):
+    current_span = get_current_span()
+    if not current_span:
+        return
+
+    stmt_obj = context.compiled.statement if context.compiled else None
```
Comment (tyacbovi) [inline rest-service/manager_rest/storage/storage_manager.py]: Maybe add an explanation or a ref to some doc

File: rest-service/manager_rest/test/base_test.py [hunk 37]
```diff
@@ -339,6 +341,12 @@ def create_configuration(self):
         test_config.authorization_permissions = auth_dict['permissions']
         test_config.security_encryption_key = \
             'f2ytTjQ-R2yKFMzgqDAw6vgQIHGZ9SiJoW-BhktapFQ='
+        if getattr(self, 'setup_config_enable_tracing', None) is not None:
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/base_test.py]: you can init these variables with there defaults, so this will be more cleaner

File: rest-service/manager_rest/test/base_test.py [hunk 38]
```diff
@@ -732,3 +740,24 @@ def _test_invalid_input(self, func, argument, *args):
...
             *args
         )
+
+    def _patch_jaeger_config(self):
+        config.j_client.Config = MagicMock()
+        self.jaeger_mock_config = config.j_client.Config
+        self.jaeger_mock_config.return_value = self.jaeger_mock_config
+
+
+class WithInitTracerTestCase(BaseServerTestCase):
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/base_test.py]: TracerTestCase maybe it's a better name

File: rest-service/manager_rest/test/endpoints/test_app.py [hunk 39]
```diff
@@ -29,3 +32,56 @@ def test_get_root_404(self):
...
+
+    def test_does_not_init_tracer(self):
+        # Dummy call to invoke the tracer initialization if enabled.
+        self.app.get('/')
+        self.jaeger_mock_config.assert_not_called()
+
+
+@attr(client_min_version=base_test.LATEST_API_VERSION,
+      client_max_version=base_test.LATEST_API_VERSION,
+      setup_config_enable_tracing=True,
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_app.py]: use test fixture of pytest, not this.
attr is for running a test with a filter specified here

File: rest-service/manager_rest/test/endpoints/test_utils.py [hunk 40]
```diff
@@ -95,6 +97,57 @@ def mock_get_platform():
...
+
+    @patch('manager_rest.utils.span_in_context')
+    @patch('manager_rest.utils.get_current_span')
+    def test_tracing_wrapper_works(self, curr_span, span_ctx):
+        curr_span.return_value = curr_span
+        current_app.tracer.start_span.reset_mock()
+
+        dummy_f, decorated_f, return_value, _ = get_dummy_function_and_params(
+            'not_foo')
+        r = decorated_f('bar', k='v')
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_utils.py]: r is not a good name :)

File: rest-service/manager_rest/test/endpoints/test_utils.py [hunk 41]
```diff
@@ -126,3 +179,18 @@ def print_progress(read, total):
...
+
+    :param func_trace_name: function name to pass to the traces() decorator.
+    :return: (dummy function, decorated dummy function, return value,
+     function name)
+    """
+    decorated_func_name = 'foo'
+    return_value = 'something'
+    foo = MagicMock(return_value=return_value)
+    foo.__name__ = decorated_func_name
+    f = traces(func_trace_name)(foo)
```
Comment (tyacbovi) [inline rest-service/manager_rest/test/endpoints/test_utils.py]: maybe open the option to decorate the func with others
